sql/schema.go

```diff
@@ -141,27 +153,71 @@ func (s *Schema) Migrate(logger *zap.Logger, db Database, before, vacuumState in
 					return fmt.Errorf("apply %s: %w", m.Name(), err)
 				}
 			}
-			// version is set intentionally even if actual migration was skipped
-			if _, err := tx.Exec(fmt.Sprintf("PRAGMA user_version = %d;", m.Order()), nil, nil); err != nil {
-				return fmt.Errorf("update user_version to %d: %w", m.Order(), err)
+			if err := s.setVersion(tx, m.Order()); err != nil {
+				return err
 			}
 			return nil
 		}); err != nil {
-			err = errors.Join(err, db.Close())
-			return err
+			return errors.Join(err, db.Close())
 		}
 
 		if vacuumState != 0 && before <= vacuumState {
 			if err := Vacuum(db); err != nil {
-				err = errors.Join(err, db.Close())
-				return err
+				return errors.Join(err, db.Close())
 			}
 		}
 		before = m.Order()
 	}
 	return nil
 }
 
+// MigrateTempDB performs database migration on the temporary database.
+// It doesn't use transactions and the temporary database should be considered
+// invalid and discarded if it fails.
+// The database is switched into synchronous mode with WAL journal enabled and
+// synced after the migrations are completed before setting the database version,
+// which triggers file sync.
+func (s *Schema) MigrateTempDB(logger *zap.Logger, db Database, before int) error {
+	if logger.Core().Enabled(zap.DebugLevel) {
+		db.Intercept("logQueries", logQueryInterceptor(logger))
+		defer db.RemoveInterceptor("logQueries")
+	}
+	v := before
+	for _, m := range s.Migrations {
+		if m.Order() <= v {
+			continue
+		}
+
+		if _, ok := s.skipMigration[m.Order()]; !ok {
+			if err := m.Apply(db, logger); err != nil {
+				return errors.Join(fmt.Errorf("apply %s: %w", m.Name(), err), db.Close())
```

Comment:
Similar as for the other two: The only caller of `MigrateTempDB` always closes `db`, so we don't need to close it here and use `errors.Join`.

Reply:
Fixed